components/eamxx/src/share/field/field_impl.hpp

```diff
@@ -5,9 +5,86 @@
 #include "share/util/scream_array_utils.hpp"
 #include "share/util/scream_universal_constants.hpp"
 
+#include <ekat/ekat_type_traits.hpp>
+
 namespace scream
 {
 
+template<typename ViewT,
+         typename = typename std::enable_if<Kokkos::is_view<ViewT>::value>::type>
+Field::
+Field (const identifier_type& id,
+       const ViewT& view_d)
+ : Field(id)
+{
+  constexpr auto N = ViewT::Rank;
+  using ScalarT  = typename ViewT::traits::value_type;
+  using ExeSpace = typename ViewT::traits::execution_space;
+
+  EKAT_REQUIRE_MSG ( (std::is_same<ExeSpace,typename device_t::execution_space>::value),
+      "Error! This constructor of Field requires a view from device.\n");
+
+  EKAT_REQUIRE_MSG (id.data_type()==get_data_type<ScalarT>(),
+      "Error! Input view data type does not match what is stored in the field identifier.\n"
+      " - field name: " + id.name() + "\n"
+      " - field data type: " + e2str(id.data_type()) + "\n");
+
+  const auto& fl = id.get_layout();
+  EKAT_REQUIRE_MSG (N==fl.rank(),
+      "Error! This constructor of Field requires a device view of the correct rank.\n"
+      " - field name: " + id.name() + "\n"
+      " - field rank: " + std::to_string(fl.rank()) + "\n"
+      " - view rank : " + std::to_string(N) + "\n");
+  for (int i=0; i<(N-1); ++i) {
+    EKAT_REQUIRE_MSG (view_d.extent_int(i)==fl.dims()[i],
+        "Error! Input view has the wrong i-th extent.\n"
+        " - field name: " + id.name() + "\n"
+        " - idim: " + std::to_string(i) + "\n"
+        " - layout i-th dim: " + std::to_string(fl.dims()[i]) + "\n"
+        " - view i-th dim: " + std::to_string(view_d.extent(i)) + "\n"); 
+  }
+
+  auto& alloc_prop = m_header->get_alloc_properties();
+  if (N>0 and view_d.extent_int(N-1)!=fl.dims().back()) {
+    EKAT_REQUIRE_MSG (view_d.extent_int(N-1)>=fl.dims()[N-1],
+        "Error! Input view has the wrong last extent.\n"
+        " - field name: " + id.name() + "\n"
+        " - layout last dim: " + std::to_string(fl.dims()[N-1]) + "\n"
+        " - view last dim: " + std::to_string(view_d.extent(N-1)) + "\n"); 
+
+    // We have a padded view. We don't know what the pack size was, so we pick the largest
+    // power of 2 that divides the last extent
+    auto last_view_dim = view_d.extent_int(N-1);
+    int last_fl_dim    = fl.dims().back();
+
+    // This should get the smallest pow of 2 that gives npacks*pack_size==view_last_dim
```

Comment:
Won't this always be pow2 = 1?  Since no matter what a packsize of 2 should always fit?

Reply:
No. Say that the physical dim is 24, and the view was allocated with pack size 16, so it has a last extent 32. With `ps=2` you get
```
npacks = (24 + 2 - 1) / 2 = 12;
packed_length = 12 * 2 = 24;
```
Since `24!=36`, `ps=2` is not accepted. Everything up to 8 is rejected, since with `ps=1,2,4,8` you can fit 24 exactly, with no padding, so you would get `packed_length=24`. Only for `ps=16` you start to have padding, and you indeed get 2 packs, with `packed_length = 16 * 2 = 32`, which matches the view extent, and is therefore accepted.